kornia/filters/bilateral.py

```diff
@@ -1,13 +1,64 @@
 from __future__ import annotations
 
 from kornia.core import Module, Tensor, pad
-from kornia.core.check import KORNIA_CHECK_IS_TENSOR, KORNIA_CHECK_SHAPE
+from kornia.core.check import KORNIA_CHECK, KORNIA_CHECK_IS_TENSOR, KORNIA_CHECK_SHAPE
 
-from .gaussian import get_gaussian_kernel2d
-from .kernels import _unpack_2d_ks
+from .kernels import _unpack_2d_ks, get_gaussian_kernel2d
 from .median import _compute_zero_padding
 
 
+def _bilateral_blur(
+    input: Tensor,
+    guidance: Tensor | None,
+    kernel_size: tuple[int, int] | int,
+    sigma_color: float | Tensor,
+    sigma_space: tuple[float, float] | Tensor,
+    border_type: str = 'reflect',
+    color_distance_type: str = "l1",
+) -> Tensor:
+    "Single implementation for both Bilateral Filter and Joint Bilateral Filter"
+
+    KORNIA_CHECK_IS_TENSOR(input)
+    KORNIA_CHECK_SHAPE(input, ['B', 'C', 'H', 'W'])
+    if guidance is not None:
+        KORNIA_CHECK_IS_TENSOR(guidance)
```

Comment:
The three lines below could be reduced to input.shape==guidance.shape ?

Reply:
I want to allow guidance and input having different number of channels e.g. guidance is grayscale, input is RGB. What do you think? Should we allow that?

Reply:
I guess it will only work for guidance being num channels 1 or same as input, right ? 1 Because of Broadcasting

Reply:
`color_distance` will always have 1 channel since we doing reduction across channel dimension (i.e. `.sum(dim=1, keepdim=True)`). So it doesn't matter how many channels `guidance` has. The intensity/color kernel to be applied is the same across all channels of `input`.

Reply:
i don't have any preference here -- up to you in terms of what you think that can be useful for other to research/experiment